### PR TITLE
Implement distributed learning iterations over mini-batches

### LINKED ISSUE #4: Parameterize size of mini-batches
Parameterize size of mini-batches so that distributed learning and weights aggregating phases can be optimized.

### PR DESCRIPTION
**Pull request by bowni**
Implements and closes #4:
- for each epoch, splits the nodes datasets into mini-batches
- iterates over mini-batches for training individual models and aggregating them

### CODE REVIEW

File: fl_training.py [hunk 1]
```diff
@@ -149,191 +194,169 @@ def compute_test_score(
...
                 history = node_model.fit(
-                    node.x_train,
-                    node.y_train,
+                    minibatched_x_train[node_index][minibatch+1],
+                    minibatched_y_train[node_index][minibatch+1],
                     batch_size=constants.BATCH_SIZE,
                     epochs=1,
                     verbose=0,
-                    validation_data=(node.x_val, node.y_val),
+                    validation_data=(minibatched_x_val[node_index][minibatch+1], minibatched_y_val[node_index][minibatch+1]),
```
Comment (Thomas-Galtier) [inline fl_training.py]: I not sure but should we rather keep the hole validation data here?

File: fl_training.py [hunk 2]
```diff
@@ -149,191 +194,169 @@ def compute_test_score(
...
+        minibatched_x_train = [None] * nodes_count
+        minibatched_x_val = [None] * nodes_count
+        minibatched_y_train = [None] * nodes_count
+        minibatched_y_val = [None] * nodes_count
+        split_indices = np.arange(minibatch_count+1) / minibatch_count
+        for node_index, node in enumerate(node_list):
+            minibatched_x_train[node_index] = np.split(node.x_train, (split_indices * len(node.x_train)).astype(int))
+            minibatched_x_val[node_index] = np.split(node.x_val, (split_indices * len(node.x_val)).astype(int))
+            minibatched_y_train[node_index] = np.split(node.y_train, (split_indices * len(node.y_train)).astype(int))
+            minibatched_y_val[node_index] = np.split(node.y_val, (split_indices * len(node.y_val)).astype(int))
```
Comment (Thomas-Galtier) [inline fl_training.py]: I not sure but should we rather keep the whole validation data here?

File: fl_training.py (hunk 1 above)
Comment (bowni) [reply fl_training.py]: By the whole validation dataset you mean the early stopping validation dataset? If we do that don't we risk some kind of "early stopping overfitting"? cc @RomainGoussault

**Comment by bowni** [comment]
I just rebased with `master` (following recent merge of PR #80) to enable an easier review, test and merge @RomainGoussault.

File: fl_training.py (hunk 1 above)
Comment (Thomas-Galtier) [reply fl_training.py]: I don't think so, because the validation set is not used during the fit. [see here](https://www.tensorflow.org/api_docs/python/tf/keras/Model )The validation set is put in argument only to track the loss after each epoch.

File: scenario.py [hunk 3]
```diff
@@ -60,11 +60,19 @@ def __init__(self, params, experiment_path):
...
+            self.aggregation_weighting = params["aggregation_weighting"]
+        else:
+            self.aggregation_weighting = "uniform"
+
         self.node_list = []
 
         self.contributivity_list = []
 
         self.epoch_count = 40
+        self.minibatch_count = 20
```
Comment (RomainGoussault) [inline scenario.py]: Specify value in config file so we can easily launch experiments with different values of minibatch.

Comment (RomainGoussault) [inline scenario.py]: I would also rename to `minibatch_size `(or even to `federated_learning_batch_size`) as the minibatch_size would generally be greater than the batch_size !

Comment (RomainGoussault) [inline scenario.py]: I was wondering if we should use ratio or absolute values but I think absolute values is better.
Also we probably need an easy way to specify to not use mini batch at all.

File: fl_training.py (hunk 2 above)
Comment (RomainGoussault) [inline fl_training.py]: A bit hard to follow.
I propose: `minibatched_x_train[node_index] = np.split(node.x_train, minibatches_count)`

File: fl_training.py [hunk 4]
```diff
@@ -149,191 +194,169 @@ def compute_test_score(
...
+
+        # Iterate over mini-batches for training and aggregation
+        for minibatch in range(minibatch_count):
```
Comment (RomainGoussault) [inline fl_training.py]: `minibatch_index `instead of `minibatch`

File: fl_training.py [hunk 5]
```diff
@@ -149,191 +194,169 @@ def compute_test_score(
...
-                        metrics=["accuracy"],
-                    )
+                print("   Training on node " + str(node_index) + " - " + str(node))
 
-                # Train on whole node local data set
+                # Train on node local data set
                 history = node_model.fit(
-                    node.x_train,
-                    node.y_train,
+                    minibatched_x_train[node_index][minibatch+1],
```
Comment (RomainGoussault) [inline fl_training.py]: Why `+1`?

File: fl_training.py (hunk 1 above)
Comment (RomainGoussault) [reply fl_training.py]: Good question !!
I think using the whole validation data set would make more sense now, as we only use it for plotting for now

File: fl_training.py [hunk 6]
```diff
@@ -141,6 +143,49 @@ def compute_test_score_with_scenario(scenario, is_save_fig=False):
...
 #%% Distributed learning training
 
+def prepare_aggregation_weights(aggregation_weighting, nodes_count, node_list):
+    """Returns a list of weights for the weighted average aggregation of model weights"""
+
+    aggregation_weights = []
+    if aggregation_weighting == "uniform":
+        aggregation_weights = [1] * nodes_count
+    elif aggregation_weighting == "data-volume":
+        aggregation_weights = [len(node.x_train) for node in node_list]
```
[the remark this answers is not in the document]
Comment (Thomas-Galtier) [reply fl_training.py]: actually the average method normalizes the weights by default so it is not needed

[the remark this answers is not in the document]
Comment (bowni) [reply fl_training.py]: @RomainGoussault when a change is done on another branch and merged on `master`, I think it is cleaner to not re-do it in the new branch you started before that merge. The future merge will handle that.
Although I thought I rebased properly, so now this looks weird. I'll check

File: fl_training.py [hunk 7]
```diff
@@ -149,191 +194,169 @@ def compute_test_score(
     y_val_global,
     x_test,
     y_test,
+    aggregation_weighting="uniform",
+    minibatch_count=20,
```
[the remark this answers is not in the document]
Comment (bowni) [reply fl_training.py]: Yes we do need it here. This method is not always used with a scenario perspective, it can also be used independently to compute the distributed learning on any set of nodes.

File: scenario.py [hunk 8]
```diff
@@ -51,7 +51,7 @@ def __init__(self, params, experiment_path):
         if "corrupted_nodes" in params:
             self.corrupted_nodes = params["corrupted_nodes"]
         else:
-            self.corrupted_nodes = ["not_corrupted"] * self.nodes_count
+            self.corrupted_nodes = ["not-corrupted"] * self.nodes_count
```
[the remark this answers is not in the document]
Comment (bowni) [reply scenario.py]: Hmmm, this was a way to differentiate variable names (with underscores) and names of variable values (with dashes), of which I am not at the origin. I can change it, but please note it implies changes in the code for corrupting datasets.

File: scenario.py (hunk 3 above)
Comment (bowni) [reply scenario.py]: Ok, will do

Comment (bowni) [reply scenario.py]: Hmm, this I am not a fan. This variable is not for a "size", it is for a "count".

File: fl_training.py (hunk 2 above)
Comment (bowni) [reply fl_training.py]: I did that in the beginning :-) But it doesn't work, `np.split()` can't accept an integer if it doesn't divide perfectly. So the proper way, as documented in NumPy doc, is to pass indices where to split.

File: fl_training.py (hunk 5 above)
Comment (bowni) [reply fl_training.py]: Because the first piece of the split is empty, by construction.

File: contributivity_measures.py [hunk 9]
```diff
@@ -53,7 +53,7 @@ def compute_independent_scores(
 #%% Generalization of Shapley Value computation
 
 
-def compute_SV(node_list, epoch_count, x_val_global, y_val_global, x_test, y_test):
+def compute_SV(node_list, epoch_count, x_val_global, y_val_global, x_test, y_test, aggregation_weighting="uniform", minibatch_count=20):
```
[the remark this answers is not in the document]
Comment (bowni) [reply contributivity_measures.py]: Ok done

File: fl_training.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bowni) [reply fl_training.py]: Ok, fixed

File: scenario.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bowni) [reply scenario.py]: Ok, done

File: scenario.py (hunk 3 above)
Comment (bowni) [reply scenario.py]: Ok, done

File: fl_training.py (hunk 4 above)
Comment (bowni) [reply fl_training.py]: Ok, done

File: fl_training.py (hunk 5 above)
Comment (bowni) [reply fl_training.py]: Ok, clarified

File: fl_training.py (hunk 1 above)
Comment (bowni) [reply fl_training.py]: Ok, done

File: scenario.py (hunk 3 above)
Comment (bowni) [reply scenario.py]: I thought about that: I am not sure. Actually, our distributed learning algo doesn't really make sense without the iteration over mini-batches. Iterating only over epochs is really too simplistic, and I don't think we want a parameter to enable the user to do it.

**Comment by bowni** [comment]
To do:

- [ ] Assert that `minibatch_count` < `min(data samples of smaller node)`

File: fl_training.py [hunk 10]
```diff
@@ -150,198 +211,146 @@ def compute_test_score(
...
-                        optimizer="adam",
-                        metrics=["accuracy"],
-                    )
+            # Iterate over nodes for training each individual model
+            for node_index, node in enumerate(node_list):
 
-                # Train on whole node local data set
+                # Train on node local data set
+                print("         Training on node " + str(node_index) + " - " + str(node))
                 history = node_model.fit(
```
Comment (RomainGoussault) [inline fl_training.py:270]: `node_model` is not cleared between 2 nodes. It means node_0 trains. Then node_1 trains sequentially on the model trained on node_0. Then node_2 trains sequentially on the model trained on node_0 and node_1 etc.
I think it is valid FL scenario (even probably better than what we did before !) but different to the federated averaging where every node communicates only with a central server/aggregator.

Comment (RomainGoussault) [reply fl_training.py:270]: In fact we now implemented Sequential FL instead of previous Parallel implementation:
image
(taken from Substra whitepaper)

**Comment by bowni** [comment]
OK @RomainGoussault, I wasn't familiar with this, shame on me!
So I took the following approach: for a given mini-batch, before iterating over all nodes, build a new model for each node. How does that look?

**Comment by bowni** [comment]
Just rebased with recent merge on `master`.
